articles/realtime.md

```diff
@@ -0,0 +1,293 @@
+* In practice, only works for small processes
+
+### Custom fixed allocators for STL containers
+
+An alternative to the above approach is to implement custom allocators for STL containers
+that only allocate memory on the stack.
+
+There are various comprehensive tutorials
+[already written](http://www.codeguru.com/cpp/article.php/c18503/C-Programming-Stack-Allocators-for-STL-Containers.htm)
+for this task.
+
+Pros:
+* Use existing STL code with deterministic computation and without allocating global dynamic memory pool
+* More modular solution
+
+Cons:
+* Complex to implement
+
+### Global variables and (static) arrays
+
+Global variables are preallocated at the start of a process, thus assigning and accessing them
+is "real-time" safe. However, this strategy comes with the many disadvantages of using global
+variables.
+
+### Use inheritance sparingly
+
+Classes with many levels of inheritance may not be real-time safe because of vtable overhead access.
+When executing an inherited function, the program needs to access the data used in the function,
+the vtable for the class, and the instructions for the function, which are all stored in different
+parts of memory, and may or may not be stored in cache together.
+
+In general, C++ patterns with poor cache locality are not well-suited to real-time environments.
+Another such pattern is the opaque pointer idiom (PIMPL), which is convenient for ABI compatibility
+and speeding up compile times. However, bouncing between the memory location for the object and its
+private data pointer causes the cache to "spill" as it loads one chunk of memory and then another,
+unrelated chunk for almost every function in the PIMPLized object.
+
+### Exceptions
+
+Throwing an exception can put large objects on the stack, which is undesirable in real-time
+programming since we cannot allocate memory on the heap. On modern C++ compilers, catching an
+exception has little memory or time overhead, but can lead to unexpected code growth.
+
+## Device I/O
+Interacting with physical devices (disk I/O, printing to the screen, etc.) may introduce
+unacceptable latency in the real-time code path, since the process is often forced to wait
+on slow physical phenomena. Additionally, many I/O calls such as `fopen` result in pagefaults.
+
+Keep disk reads/writes at the beginning or end of the program, outside of the RT code path.
+
+Spin up userspace threads (executing on a different CPU from the RT code) to print output to
+the screen.
+
+## Multithreaded Programming and Synchronization
+
+Real-time computation requirements change the typical paradigm of multithreaded programming.
+Program execution may not block asynchronously, and threads must be scheduled deterministically.
+A real-time operating system will fulfill this scheduling requirement, but there are still pitfalls
+for the developer to fall into. Typical assumptions about locking critical sections will fall apart
+if thread execution can be arbitrarily preempted.
```

Comment:
Is it referring to priority inversion? Because even in non real time systems the thread execution can be arbitrarily preempted.